pkg/expression/constant.go

```diff
@@ -139,26 +137,25 @@ func (d *ParamMarker) GetUserVar(ctx EvalContext) types.Datum {
 	return ctx.GetParamValue(d.order)
 }
 
-func (d *ParamMarker) getUserVarWithInternalCtx() types.Datum {
-	// TODO: remove this function in the future
-	sessionVars := d.ctx.GetSessionVars()
-	return sessionVars.PlanCacheParams.GetParamValue(d.order)
-}
-
-// String implements fmt.Stringer interface.
-func (c *Constant) String() string {
+// StringWithCtx implements Expression interface.
+func (c *Constant) StringWithCtx(ctx EvalContext) string {
 	if c.ParamMarker != nil {
-		dt := c.ParamMarker.getUserVarWithInternalCtx()
+		intest.Assert(ctx != nil, "context is nil for ParamMarker")
+		if ctx == nil {
+			return "?"
```

Comment:
How about triggering a panic here? Then if someone forget to pass `ctx`, it can be exposed immediately.

Reply:
As a result after discussion, we agreed that `panic` here directly will have big risk, so only return a `"?"` and have an assertion for test is good.

Later (not in this PR), I'll try to consider how to extend the tests of binary protocol, because most of the current tests didn't have `param`.